From e9264df6e63b5df87d81e950675df7290ad43615 Mon Sep 17 00:00:00 2001
From: Andrew Eikum <aeikum@codeweavers.com>
Date: Wed, 15 Jul 2020 14:57:28 -0500
Subject: [PATCH] wineboot: On prefix upgrade, update win10 build number

Some games (Death Stranding) require later build numbers than we had
shipped earlier. So fix it up on existing prefixes.
---
 programs/wineboot/wineboot.c | 38 ++++++++++++++++++++++++++++++++++++
 1 file changed, 38 insertions(+)

diff --git a/programs/wineboot/wineboot.c b/programs/wineboot/wineboot.c
index 569db5ee94d..a0a72040843 100644
--- a/programs/wineboot/wineboot.c
+++ b/programs/wineboot/wineboot.c
@@ -1451,6 +1451,43 @@ static void update_user_profile(void)
     LocalFree(sid);
 }

+static void update_win_version(void)
+{
+    static const WCHAR win10_buildW[] = L"17763";
+
+    HKEY cv_h;
+    DWORD type, sz;
+    WCHAR current_version[256];
+
+    if(RegOpenKeyExW(HKEY_LOCAL_MACHINE, L"Software\\Microsoft\\Windows NT\\CurrentVersion",
+                0, KEY_ALL_ACCESS, &cv_h) == ERROR_SUCCESS){
+        /* get the current windows version */
+        sz = sizeof(current_version);
+        if(RegQueryValueExW(cv_h, L"CurrentVersion", NULL, &type, (BYTE *)current_version, &sz) == ERROR_SUCCESS &&
+                type == REG_SZ){
+            if(!wcscmp(current_version, L"10.0")){
+                RegSetValueExW(cv_h, L"CurrentBuild", 0, REG_SZ, (const BYTE *)win10_buildW, sizeof(win10_buildW));
+                RegSetValueExW(cv_h, L"CurrentBuildNumber", 0, REG_SZ, (const BYTE *)win10_buildW, sizeof(win10_buildW));
+            }
+        }
+        RegCloseKey(cv_h);
+    }
+
+    if(RegOpenKeyExW(HKEY_LOCAL_MACHINE, L"Software\\Wow6432Node\\Microsoft\\Windows NT\\CurrentVersion",
+                0, KEY_ALL_ACCESS, &cv_h) == ERROR_SUCCESS){
+        /* get the current windows version */
+        sz = sizeof(current_version);
+        if(RegQueryValueExW(cv_h, L"CurrentVersion", NULL, &type, (BYTE *)current_version, &sz) == ERROR_SUCCESS &&
+                type == REG_SZ){
+            if(!wcscmp(current_version, L"10.0")){
+                RegSetValueExW(cv_h, L"CurrentBuild", 0, REG_SZ, (const BYTE *)win10_buildW, sizeof(win10_buildW));
+                RegSetValueExW(cv_h, L"CurrentBuildNumber", 0, REG_SZ, (const BYTE *)win10_buildW, sizeof(win10_buildW));
+            }
+        }
+        RegCloseKey(cv_h);
+    }
+}
+
 /* execute rundll32 on the wine.inf file if necessary */
 static void update_wineprefix( BOOL force )
 {
@@ -1496,6 +1533,7 @@ static void update_wineprefix( BOOL force )
         }
         install_root_pnp_devices();
         update_user_profile();
+        update_win_version();

         WINE_MESSAGE( "wine: configuration in %s has been updated.\n", debugstr_w(prettyprint_configdir()) );
     }
From e3f3c07144cda0c4aa25a3d104ca76c17e36bfdf Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Tue, 14 Sep 2021 23:03:54 +0300
Subject: [PATCH] Bump current build number to 18363 (Win10 1909).

CW-Bug-ID: #19427

For DeathLoop.
---
 dlls/kernel32/version.rc        | 8 ++++----
 dlls/kernelbase/version.c       | 2 +-
 dlls/ntdll/unix/signal_x86_64.c | 4 ++--
 loader/wine.inf.in              | 8 ++++----
 programs/wineboot/wineboot.c    | 2 +-
 5 files changed, 12 insertions(+), 12 deletions(-)

diff --git a/dlls/kernel32/version.rc b/dlls/kernel32/version.rc
index b6002f51f7a..c58acb8a8b6 100644
--- a/dlls/kernel32/version.rc
+++ b/dlls/kernel32/version.rc
@@ -26,9 +26,9 @@ LANGUAGE LANG_ENGLISH, SUBLANG_DEFAULT
 #define WINE_FILENAME_STR "kernel32.dll"

 /* these values come from Windows 10 Version 1909 */
-#define WINE_FILEVERSION 10,0,18362,1350
-#define WINE_FILEVERSION_STR "10.0.18362.1350"
-#define WINE_PRODUCTVERSION 10,0,18362,1350
-#define WINE_PRODUCTVERSION_STR "10.0.18362.1350"
+#define WINE_FILEVERSION 10,0,18363,900
+#define WINE_FILEVERSION_STR "10.0.18363.900"
+#define WINE_PRODUCTVERSION 10,0,18363,900
+#define WINE_PRODUCTVERSION_STR "10.0.18363.900"

 #include "wine/wine_common_ver.rc"
diff --git a/dlls/kernelbase/version.c b/dlls/kernelbase/version.c
index bd878cb324d..febb78d94e9 100644
--- a/dlls/kernelbase/version.c
+++ b/dlls/kernelbase/version.c
@@ -151,7 +151,7 @@ static const struct
     },
     /* Windows 10 */
     {
-        { 10, 0, 18362 },
+        { 10, 0, 18363 },
         {0x8e0f7a12,0xbfb3,0x4fe8,{0xb9,0xa5,0x48,0xfd,0x50,0xa1,0x5a,0x9a}}
     }
 };
diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 4853db16e77..faba4be0b91 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -2448,11 +2448,11 @@ HKLM,"System\CurrentControlSet\Services\Winsock\Parameters",,16
 HKLM,"System\CurrentControlSet\Services\Winsock2\Parameters\Protocol_Catalog9\Catalog_Entries",,16

 [VersionInfo]
-HKLM,%CurrentVersionNT%,"CurrentVersion",2,"6.3"
+HKLM,%CurrentVersionNT%,"CurrentVersion",2,"10.0"
 HKLM,%CurrentVersionNT%,"CurrentMajorVersionNumber",0x10001,10
 HKLM,%CurrentVersionNT%,"CurrentMinorVersionNumber",0x10001,0
-HKLM,%CurrentVersionNT%,"CurrentBuild",2,"18362"
-HKLM,%CurrentVersionNT%,"CurrentBuildNumber",2,"18362"
+HKLM,%CurrentVersionNT%,"CurrentBuild",2,"18363"
+HKLM,%CurrentVersionNT%,"CurrentBuildNumber",2,"18363"
 HKLM,%CurrentVersionNT%,"CurrentType",2,"Multiprocessor Free"
 HKLM,%CurrentVersionNT%,"DigitalProductId",1,00,00,00,00,00,00,00,00,00,00,00,\
 00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,\
diff --git a/programs/wineboot/wineboot.c b/programs/wineboot/wineboot.c
index f78d4bdae25..eb69a5941a9 100644
--- a/programs/wineboot/wineboot.c
+++ b/programs/wineboot/wineboot.c
@@ -1706,7 +1706,7 @@ static void update_user_profile(void)

 static void update_win_version(void)
 {
-    static const WCHAR win10_buildW[] = L"17763";
+    static const WCHAR win10_buildW[] = L"18363";

     HKEY cv_h;
     DWORD type, sz;
From 7381437f853f4a1706ff80aa74860fc5366f2fd0 Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Wed, 29 Sep 2021 14:39:06 +0300
Subject: [PATCH] wine.inf: Add ReleaseId value to %CurrentVersionNT%.

CW-Bug-Id: #19484
---
 loader/wine.inf.in | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 2db2abab2d8..1cfe19dace6 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -2423,6 +2423,7 @@ HKLM,%CurrentVersionNT%,"DigitalProductId",1,00,00,00,00,00,00,00,00,00,00,00,\
 00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,\
 00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00
 HKLM,%CurrentVersionNT%,"ProductName",2,"Windows 10 Pro"
+HKLM,%CurrentVersionNT%,"ReleaseId",,"1909"
 HKLM,%Control%\ProductOptions,"ProductType",2,"WinNT"
 HKLM,%Control%\Windows,"CSDVersion",0x10003,0
 HKLM,%Control%\Session Manager\Environment,"OS",2,"Windows_NT"
From 6de51bc22e5bd220ba79cbfae052b08abc4289d2 Mon Sep 17 00:00:00 2001
From: Paul Gofman <gofmanp@gmail.com>
Date: Fri, 14 Jan 2022 11:46:04 +0300
Subject: [PATCH] Bump current build number to 19043 (Win10 2009).

---
 dlls/kernel32/version.rc        | 10 +++++-----
 dlls/kernelbase/version.c       |  2 +-
 dlls/ntdll/unix/signal_x86_64.c |  3 ++-
 dlls/ntdll/version.c            |  2 +-
 loader/wine.inf.in              | 14 ++++++++------
 programs/wineboot/wineboot.c    |  2 +-
 6 files changed, 18 insertions(+), 15 deletions(-)

diff --git a/dlls/kernel32/version.rc b/dlls/kernel32/version.rc
index c58acb8a8b6..f91262dfa8e 100644
--- a/dlls/kernel32/version.rc
+++ b/dlls/kernel32/version.rc
@@ -25,10 +25,10 @@ LANGUAGE LANG_ENGLISH, SUBLANG_DEFAULT
 #define WINE_FILEDESCRIPTION_STR "Wine kernel DLL"
 #define WINE_FILENAME_STR "kernel32.dll"

-/* these values come from Windows 10 Version 1909 */
-#define WINE_FILEVERSION 10,0,18363,900
-#define WINE_FILEVERSION_STR "10.0.18363.900"
-#define WINE_PRODUCTVERSION 10,0,18363,900
-#define WINE_PRODUCTVERSION_STR "10.0.18363.900"
+/* these values come from Windows 10 Version 2009 */
+#define WINE_FILEVERSION 10,0,19043,1466
+#define WINE_FILEVERSION_STR "10.0.19043.1466"
+#define WINE_PRODUCTVERSION 10,0,19043,1466
+#define WINE_PRODUCTVERSION_STR "10.0.19043.1466"

 #include "wine/wine_common_ver.rc"
diff --git a/dlls/kernelbase/version.c b/dlls/kernelbase/version.c
index b4f9406f3c6..ab4ec32a260 100644
--- a/dlls/kernelbase/version.c
+++ b/dlls/kernelbase/version.c
@@ -158,7 +158,7 @@ static const struct
     },
     /* Windows 10 */
     {
-        { 10, 0, 18363 },
+        { 10, 0, 19043 },
         {0x8e0f7a12,0xbfb3,0x4fe8,{0xb9,0xa5,0x48,0xfd,0x50,0xa1,0x5a,0x9a}}
     }
 };
diff --git a/dlls/ntdll/version.c b/dlls/ntdll/version.c
index 077ab65a502..6c846d0ac33 100644
--- a/dlls/ntdll/version.c
+++ b/dlls/ntdll/version.c
@@ -167,7 +167,7 @@ static const RTL_OSVERSIONINFOEXW VersionData[NB_WINDOWS_VERSIONS] =
     },
     /* WIN10 */
     {
-        sizeof(RTL_OSVERSIONINFOEXW), 10, 0, 18362, VER_PLATFORM_WIN32_NT,
+        sizeof(RTL_OSVERSIONINFOEXW), 10, 0, 19043, VER_PLATFORM_WIN32_NT,
         L"", 0, 0, VER_SUITE_SINGLEUSERTS, VER_NT_WORKSTATION, 0
     },

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 769ff984187..0a41810d884 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -2456,8 +2456,8 @@ HKLM,"System\CurrentControlSet\Services\Winsock2\Parameters\Protocol_Catalog9\Ca
 HKLM,%CurrentVersionNT%,"CurrentVersion",2,"10.0"
 HKLM,%CurrentVersionNT%,"CurrentMajorVersionNumber",0x10001,10
 HKLM,%CurrentVersionNT%,"CurrentMinorVersionNumber",0x10001,0
-HKLM,%CurrentVersionNT%,"CurrentBuild",2,"18363"
-HKLM,%CurrentVersionNT%,"CurrentBuildNumber",2,"18363"
+HKLM,%CurrentVersionNT%,"CurrentBuild",2,"19043"
+HKLM,%CurrentVersionNT%,"CurrentBuildNumber",2,"19043"
 HKLM,%CurrentVersionNT%,"CurrentType",2,"Multiprocessor Free"
 HKLM,%CurrentVersionNT%,"DigitalProductId",1,00,00,00,00,00,00,00,00,00,00,00,\
 00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,\
@@ -2466,8 +2466,9 @@ HKLM,"System\CurrentControlSet\Services\Winsock2\Parameters\Protocol_Catalog9\Ca
 00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,\
 00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,\
 00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%,"DisplayVersion",2,"21H1"
 HKLM,%CurrentVersionNT%,"ProductName",2,"Windows 10 Pro"
-HKLM,%CurrentVersionNT%,"ReleaseId",,"1909"
+HKLM,%CurrentVersionNT%,"ReleaseId",,"2009"
 HKLM,%Control%\ProductOptions,"ProductType",2,"WinNT"
 HKLM,%Control%\Windows,"CSDVersion",0x10003,0
 HKLM,%Control%\Session Manager\Environment,"OS",2,"Windows_NT"
diff --git a/programs/wineboot/wineboot.c b/programs/wineboot/wineboot.c
index 14d12883813..5d1190583b1 100644
--- a/programs/wineboot/wineboot.c
+++ b/programs/wineboot/wineboot.c
@@ -1415,7 +1415,7 @@ static void update_user_profile(void)

 static void update_win_version(void)
 {
-    static const WCHAR win10_buildW[] = L"18363";
+    static const WCHAR win10_buildW[] = L"19043";

     HKEY cv_h;
     DWORD type, sz;
From fc3d95a311a8e329af3d55a88fc68b41869af863 Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Wed, 22 Dec 2021 00:57:25 +0300
Subject: [PATCH] wine.inf: Set a valid Win10 ProductId.

CW-Bug-Id: #19702
---
 loader/wine.inf.in | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 0a41810d884..1c2286d722f 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -549,7 +549,7 @@ HKCU,%CurrentVersion%\Run,,16
 HKCU,%CurrentVersionNT%\Winlogon,,16
 HKLM,%CurrentVersion%,"CommonFilesDir",,"%16427%"
 HKLM,%CurrentVersion%,"FirstInstallDateTime",1,21,81,7c,23
-HKLM,%CurrentVersion%,"ProductId",,"12345-oem-0000001-54321"
+HKLM,%CurrentVersion%,"ProductId",,"00330-50000-00000-AAOEM"
 HKLM,%CurrentVersion%,"ProgramFilesDir",,"%16422%"
 HKLM,%CurrentVersion%,"ProgramFilesPath",0x20000,"%%ProgramFiles%%"
 HKLM,%CurrentVersion%,"RegisteredOrganization",2,""
@@ -573,7 +573,7 @@ HKLM,%CurrentVersion%\Setup\WindowsFeatures\WindowsMediaVersion,,,"12.0.7601.188
 HKLM,%CurrentVersion%\Shell Extensions\Approved,,16
 HKLM,%CurrentVersion%\Uninstall,,16
 HKLM,%CurrentVersionNT%,"InstallDate",0x10003,1273299354
-HKLM,%CurrentVersionNT%,"ProductId",,"12345-oem-0000001-54321"
+HKLM,%CurrentVersionNT%,"ProductId",,"00330-50000-00000-AAOEM"
 HKLM,%CurrentVersionNT%,"RegisteredOrganization",2,""
 HKLM,%CurrentVersionNT%,"RegisteredOwner",2,""
 HKLM,%CurrentVersionNT%,"SystemRoot",,"%10%"
From 80d7c0c041636e900589e8ae41c5fdd3a0470b29 Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Wed, 22 Dec 2021 00:59:32 +0300
Subject: [PATCH] wineboot: Generate better DigitalProductId.

CW-Bug-Id: #19702
---
 loader/wine.inf.in           |  4 +--
 programs/wineboot/wineboot.c | 48 ++++++++++++++++++++++++++++++++++++
 2 files changed, 50 insertions(+), 2 deletions(-)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 1c2286d722f..677ae91a406 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -2456,7 +2456,7 @@ HKLM,"System\CurrentControlSet\Services\Winsock2\Parameters\Protocol_Catalog9\Ca
 HKLM,%CurrentVersionNT%,"CurrentBuild",2,"19043"
 HKLM,%CurrentVersionNT%,"CurrentBuildNumber",2,"19043"
 HKLM,%CurrentVersionNT%,"CurrentType",2,"Multiprocessor Free"
-HKLM,%CurrentVersionNT%,"DigitalProductId",1,00,00,00,00,00,00,00,00,00,00,00,\
+HKLM,%CurrentVersionNT%,"DigitalProductId",2,00,00,00,00,00,00,00,00,00,00,00,\
 00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,\
 00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,\
 00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,\
diff --git a/programs/wineboot/wineboot.c b/programs/wineboot/wineboot.c
index 5d1190583b1..16b44ca28ab 100644
--- a/programs/wineboot/wineboot.c
+++ b/programs/wineboot/wineboot.c
@@ -68,6 +68,7 @@
 #include <winternl.h>
 #include <ddk/wdm.h>
 #include <sddl.h>
+#include <ntsecapi.h>
 #include <wine/svcctl.h>
 #include <wine/asm.h>
 #include <wine/debug.h>
@@ -1605,6 +1606,52 @@ static void usage( int status )
     exit( status );
 }

+static void create_digitalproductid(void)
+{
+    BYTE digital_product_id[0xa4];
+    char product_id[256];
+    LSTATUS status;
+    unsigned int i;
+    DWORD size;
+    DWORD type;
+    HKEY key;
+
+    if ((status = RegOpenKeyExW( HKEY_LOCAL_MACHINE, L"Software\\Microsoft\\Windows NT\\CurrentVersion",
+                       0, KEY_ALL_ACCESS, &key )))
+        return;
+    size = sizeof(product_id);
+    status = RegQueryValueExA( key, "ProductId", NULL, &type, (BYTE *)product_id, &size );
+    if (status) goto done;
+    if (!size) goto done;
+    if (product_id[size - 1])
+    {
+        if (size == sizeof(product_id)) goto done;
+        product_id[size++] = 0;
+    }
+
+    if (!RegQueryValueExA( key, "DigitalProductId", NULL, &type, NULL, &size ) && size == sizeof(digital_product_id))
+    {
+        if (RegQueryValueExA( key, "DigitalProductId", NULL, &type, digital_product_id, &size ))
+            goto done;
+        for (i = 0; i < size; ++i)
+            if (digital_product_id[i]) break;
+        if (i < size) goto done;
+    }
+
+    memset( digital_product_id, 0, sizeof(digital_product_id) );
+    *(DWORD *)digital_product_id = sizeof(digital_product_id);
+    digital_product_id[4] = 3;
+    strcpy( (char *)digital_product_id + 8, product_id );
+    *(DWORD *)(digital_product_id + 0x20) = 0x0cec;
+    *(DWORD *)(digital_product_id + 0x34) = 0x0cec;
+    strcpy( (char *)digital_product_id + 0x24, "[TH] X19-99481" );
+    digital_product_id[0x42] = 8;
+    RtlGenRandom( digital_product_id + 0x38, 0x18 );
+    RegSetValueExA( key, "DigitalProductId", 0, REG_BINARY, digital_product_id, sizeof(digital_product_id) );
+done:
+    RegCloseKey( key );
+}
+
 int __cdecl main( int argc, char *argv[] )
 {
     /* First, set the current directory to SystemRoot */
@@ -1716,6 +1763,7 @@ int __cdecl main( int argc, char *argv[] )
     }
     if (init || update) update_wineprefix( update );

+    create_digitalproductid();
     create_volatile_environment_registry_key();

     ProcessRunKeys( HKEY_LOCAL_MACHINE, L"RunOnce", TRUE, TRUE );
